astropy/coordinates/sky_coordinate.py

```diff
@@ -65,19 +66,32 @@ def _repr_data(self):
         return repr_data
 
     def _represent_as_dict(self):
-        obj = self._parent
-        attrs = (list(obj.representation_component_names) +
-                 list(frame_transform_graph.frame_attributes.keys()))
-
-        # Don't output distance if it is all unitless 1.0
-        if 'distance' in attrs and np.all(obj.distance == 1.0):
-            attrs.remove('distance')
+        sc = self._parent
+        attrs = list(sc.representation_component_names)
+
+        # Don't output distance unless it's actually distance.
+        if issubclass(sc.data.__class__, UnitSphericalRepresentation):
+            attrs = attrs[:-1]
+
+        diff = sc.data.differentials.get('s')
+        if diff is not None:
+            diff_attrs = list(sc.get_representation_component_names('s'))
+            # Don't output proper motions if they haven't been specified.
+            if issubclass(diff.base_representation, RadialRepresentation):
+                diff_attrs = diff_attrs[2:]
```

Comment:
Again this makes assumptions about attribute ordering based on the base class. Maybe a better strategy is to start by making a dict and then delete by name?

Reply:
Removing distance and radial velocity by name should be quite simple, but the names of the proper motion components can depend on both the frame (`'pm_dec'` vs `'pm_b'`) and differential type (`'pm_ra_cosdec'` vs `'pm_ra'`), so explicitly writing out all possibilities doesn't seem like a very good idea. It would be better to assume that all proper motion component names start with '`pm`', but I wouldn't know if doing so would be more reliable than assuming the ordering if some custom class is used.

Reply:
Yikes, this is quite tricky! My sense is for now to go with what you have - we haven't really worked out how to do this right in `baseframe` generally either.

But looking at this again, maybe it should just be `if isinstance(diff, RadialDifferential):`?

Reply:
@taldcroft - do you agree with going with this scheme for now? The go-by-name scheme does seem similarly fragile. To me, the best fix would be to change `get_representation_component_names` up in `baseframe` itself, but that would seem out of scope for this PR.

Reply:
OK, if there is no clean in-scope solution then we can do the simple one that works for now.